rskj-core/src/main/java/org/ethereum/datasource/RocksDbDataSource.java

```diff
@@ -0,0 +1,364 @@
+import org.ethereum.util.ByteUtil;
+import org.rocksdb.*;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.*;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+import static java.lang.System.getProperty;
+
+public class RocksDbDataSource implements KeyValueDataSource {
+
+    private static final Logger logger = LoggerFactory.getLogger("db");
+    private static final Profiler profiler = ProfilerFactory.getInstance();
+    private static final PanicProcessor panicProcessor = new PanicProcessor();
+
+    private final String databaseDir;
+    private final String name;
+    private RocksDB db;
+    private boolean alive;
+
+    // The native LevelDB insert/update/delete are normally thread-safe
+    // However close operation is not thread-safe and may lead to a native crash when
+    // accessing a closed DB.
+    // The rocksdbJNI lib has a protection over accessing closed DB but it is not synchronized
+    // This ReadWriteLock still permits concurrent execution of insert/delete/update operations
+    // however blocks them on init/close/delete operations
+    private final ReadWriteLock resetDbLock = new ReentrantReadWriteLock();
+
+    public RocksDbDataSource(String name, String databaseDir) {
+        this.databaseDir = databaseDir;
+        this.name = name;
+        logger.debug("New RocksDbDataSource: {}", name);
+    }
+
+    public static KeyValueDataSource makeDataSource(Path datasourcePath) {
+        KeyValueDataSource ds = new RocksDbDataSource(datasourcePath.getFileName().toString(), datasourcePath.getParent().toString());
+        ds.init();
+        return ds;
+    }
+
+    @Override
+    public void init() {
+        resetDbLock.writeLock().lock();
+        Metric metric = profiler.start(Profiler.PROFILING_TYPE.LEVEL_DB_INIT);
+        try (Options options = new Options()) {
+            logger.debug("~> RocksDbDataSource.init(): {}", name);
+
+            if (isAlive()) {
+                return;
+            }
+
+            Objects.requireNonNull(name, "no name set to the db");
+
+            long generalSize = 10L * 1024L * 1024L;
+
+            options.setCreateIfMissing(true);
+            options.setCompressionType(CompressionType.NO_COMPRESSION);
+            options.setArenaBlockSize(generalSize);
+            options.setWriteBufferSize(generalSize);
+            // options.cacheSize(0);
+            options.setParanoidChecks(true);
+            // options.verifyChecksums(true);
+
+            try {
+                logger.debug("Opening database");
+                Path dbPath = getPathForName(name, databaseDir);
+
+                Files.createDirectories(dbPath.getParent());
+
+                logger.debug("Initializing new or existing database: '{}'", name);
+                openDb(options, dbPath);
```

Comment:
Would be interesting to have the setAlive here, on the same level of the isAlive()

Reply:
What do you mean by `on the same level of the isAlive()`